### PR TITLE
Use soma_reader in SOMADataFrame

### PR DESCRIPTION
**Pull request by eddelbuettel**
This PR updates the read method in SOMADataFrame to use the C++ layer of TileDB-SOMA.

### CODE REVIEW

File: apis/r/R/SOMADataFrame.R [hunk 1]
```diff
@@ -126,52 +126,55 @@ SOMADataFrame <- R6::R6Class(
     #' @description Read
     #' Read a user-defined subset of data, addressed by the dataframe indexing
     #' column, and optionally filtered.
-    #' @param ids Indices specifying the rows to read.
-    #' @param column_names Character vector of column names to return.
-    #' @param value_filter A string containing a logical expression that is used
+    #' @param ids Optional indices specifying the rows to read.
```
Comment (aaronwolen) [inline apis/r/R/SOMADataFrame.R:129]: Good call spelling out the optional args

File: apis/r/R/SOMADataFrame.R [hunk 2]
```diff
@@ -126,52 +126,55 @@ SOMADataFrame <- R6::R6Class(
...
+          parsed <- do.call(what = tiledb::parse_query_condition,
+                            args = list(expr = str2lang(value_filter), ta = arr))
+          value_filter <- parsed@ptr
+      }
+      # ensure ids is a (named) list
+      if (!is.null(ids)) {
+          if (!is.list(ids))
+              ids <- list(ids)
+          if (is.null(names(ids)))
+              names(ids) <- self$dimnames() # TODO: ensure dimensionality
```
Comment (aaronwolen) [inline apis/r/R/SOMADataFrame.R:167]: I think we should require users to name lists passed to `ids` b/c we can't assume the order (or number) of elements matches the array's dimensions (unless it's a 1D array).

File: apis/r/tests/testthat/test-SOMADataFrame.R [hunk 3]
```diff
@@ -32,9 +30,13 @@ test_that("SOMADataFrame creation", {
...
-  tbl0 <- arrow::arrow_table(
-    foo = 1L:10L, bar = 1.1:10.1, baz = letters[1:10]
-  )
+  tbl0 <- arrow::arrow_table(foo = 1L:10L,
+                             bar = 1.1:10.1,
+                             baz = letters[1:10],
+                             schema=arrow::schema(arrow::field("foo", arrow::int32(), nullable=FALSE),
+                                                  arrow::field("bar", arrow::float64(), nullable=FALSE),
+                                                  arrow::field("baz", arrow::large_utf8(), nullable=FALSE)))
+
```
Comment (aaronwolen) [inline apis/r/tests/testthat/test-SOMADataFrame.R:39]: Any reason not to simplify and re-use the same schema defined above?

File: apis/r/tests/testthat/test-SOMADataFrame.R [hunk 4]
```diff
@@ -54,14 +56,46 @@ test_that("SOMADataFrame creation", {
...
+    tbl <- tibble::as_tibble(z)
+    expect_equal(nrow(tbl), 2638L)
+    expect_equal(ncol(tbl), 6L)
+
+    columns <- c("n_counts", "n_genes", "louvain")
+    sdf <- SOMADataFrame$new(uri)
+    z <- sdf$read(column_names=columns)
+    tbl <- tibble::as_tibble(z)
+    expect_equal(ncol(tbl), 3L)
+    expect_equal(colnames(tbl), columns)
```
Comment (aaronwolen) [inline apis/r/tests/testthat/test-SOMADataFrame.R:89]:
```suggestion
    tbl <- sdf$read(column_names=columns)
    expect_equal(tbl$num_columns, 3L)
    expect_equal(tbl$ColumnNames(), columns)
```

Definitely a nit but I think we could simplify here by using `arrow`'s methods directly and avoid the need to convert w/ tibble.

**Review by aaronwolen — APPROVED** [review]
A couple minor nits but overall this is awesome and a huge step forward!

File: apis/r/tests/testthat/test-SOMADataFrame.R (hunk 3 above)
Comment (eddelbuettel) [reply apis/r/tests/testthat/test-SOMADataFrame.R:39]: Yes, I showed you the error over slack.  `large_utf8` is needed.

File: apis/r/tests/testthat/test-SOMADataFrame.R (hunk 4 above)
Comment (eddelbuettel) [reply apis/r/tests/testthat/test-SOMADataFrame.R:89]: Sure but it's neither here nor there as it is in the test layer. I should probably test some content.

**Review by johnkerl — APPROVED** [review]

File: apis/r/R/SOMADataFrame.R (hunk 2 above)
Comment (eddelbuettel) [reply apis/r/R/SOMADataFrame.R]: Now done.

File: apis/r/tests/testthat/test-SOMADataFrame.R (hunk 3 above)
Comment (eddelbuettel) [reply apis/r/tests/testthat/test-SOMADataFrame.R:39]: Yup, and your follow-up helps.

File: apis/r/tests/testthat/test-SOMADataFrame.R (hunk 4 above)
Comment (eddelbuettel) [reply apis/r/tests/testthat/test-SOMADataFrame.R:88]: Done too.

**Comment by eddelbuettel** [comment]
I rebased once more and will await these CI runs and then merge if nothing unexpected bubbles up.

**Pull request merged**